src/esp/io/URDFParser.h

```diff
@@ -0,0 +1,596 @@
+        scaleShape(collision, scaleCorrection);
+      }
+    }
+    for (const auto& joint : m_joints) {
+      // scale joint offsets
+      joint.second->m_parentLinkToJointTransform.translation() *=
+          scaleCorrection;
+      // scale prismatic joint limits
+      if (joint.second->m_type == PrismaticJoint) {
+        joint.second->m_lowerLimit *= double(scaleCorrection);
+        joint.second->m_upperLimit *= double(scaleCorrection);
+      }
+    }
+
+    m_globalScaling = scaling;
+  }
+
+  //! Get the currently configured global model scaling
+  float getGlobalScaling() const { return m_globalScaling; }
+
+  /**
+   * @brief Set scaling for mass from initial values configured in URDF.
+   * Modifies various internal parameters.
+   *
+   * @param massScaling The new absolute uniform mass scale.
+   */
+  void setMassScaling(float massScaling) {
+    if (massScaling == m_massScaling) {
+      // do nothing
+      return;
+    }
+    float massScaleCorrection = massScaling / m_massScaling;
+
+    // Only need to scale the per-link mass values. These will be further
+    // processed during import.
+    for (const auto& link : m_links) {
+      Inertia& linkInertia = link.second->m_inertia;
+      linkInertia.m_mass *= double(massScaleCorrection);
+    }
+
+    m_massScaling = massScaling;
+  }
+
+  //! Get the currently configured mass scaling of the model
+  float getMassScaling() const { return m_massScaling; }
+
+ protected:
+  // scaling values which can be applied to the model after parsing
+  //! Global euclidean scaling applied to the model's transforms, asset scales,
+  //! and prismatic joint limits. Does not affect mass.
+  float m_globalScaling = 1.0;
+
+  //! Mass scaling of the model's Link inertias.
+  float m_massScaling = 1.0;
+
+  //! Scale the transformation and parameters of a Shape
+  void scaleShape(Shape& shape, float scale) {
+    shape.m_linkLocalFrame.translation() *= scale;
+    switch (shape.m_geometry.m_type) {
+      case GEOM_MESH: {
+        shape.m_geometry.m_meshScale *= scale;
+      } break;
+      case GEOM_BOX: {
+        shape.m_geometry.m_boxSize *= scale;
+      } break;
+      case GEOM_SPHERE: {
+        shape.m_geometry.m_sphereRadius *= double(scale);
+      } break;
+      case GEOM_CAPSULE:
+      case GEOM_CYLINDER: {
+        shape.m_geometry.m_capsuleRadius *= double(scale);
+        shape.m_geometry.m_capsuleHeight *= double(scale);
+      } break;
+      default:
+        break;
+    }
+  }
+};
+
+/**
+ * @brief Functional class for parsing URDF files into a URDF::Model
+ * representation.
+ */
+class Parser {
+  // URDF file path of last load call
+  std::string sourceFilePath_;
+
+  /**
+   * @brief Parse a transform into a Matrix4.
+   *
+   * @param tr The transform to fill.
+   * @param xml The source xml element to parse (e.g. <origin>).
+   * @return Success or failure.
+   */
+  bool parseTransform(Magnum::Matrix4& tr, tinyxml2::XMLElement* xml) const;
```

Comment:
Nit: it would make more sense to me if these xml elements in these function signatures were pointers-to-const instead of pointers-to-nonconst.